components/servo/message_port.rs

```diff
@@ -0,0 +1,163 @@
+                MessagePortMsg::NewTask(message_port_id, task) => {
+                    let data = match task.data {
+                        PostMessageData::StructuredClone(..) => unreachable!(),
+                        PostMessageData::Serialized(data) => data,
+                    };
+                    let _ = embedder_proxy.send(EmbedderMsg::PostMessage(
+                        webview_id,
+                        message_port_id,
+                        data,
+                    ));
+                },
+            }),
+        );
+        let first_id = MessagePortId::new();
+        let second_id = MessagePortId::new();
+        let router_id = MessagePortRouterId::new();
+        constellation_proxy.send(EmbedderToConstellationMessage::CreateEntangledMessagePorts(
+            router_id,
+            port_control_sender,
+            first_id,
+            second_id,
+        ));
+
+        let first = MessagePort::new(constellation_proxy.clone(), webview_id, first_id, second_id);
+        let second = MessagePort::new(constellation_proxy, webview_id, second_id, first_id);
+
+        (first, second)
+    }
+
+    /// Post the provided [JSValue] to the MessagePort that is entangled with this one.
+    pub fn post_message(&self, js_value: JSValue) -> Result<(), MessagePortDetached> {
+        let inner = self.0.borrow();
+        let MessagePortState::Ready {
+            entangled_id,
+            constellation_proxy,
+            ..
+        } = &*inner
+        else {
+            return Err(MessagePortDetached);
+        };
+        constellation_proxy.send(EmbedderToConstellationMessage::PostMessageToMessagePort(
+            *entangled_id,
+            js_value,
+        ));
+        Ok(())
+    }
+
+    /// Set the delegate of this MessagePort. Does nothing if this MessagePort
+    /// has already been transferred.
+    pub fn set_delegate(&self, new_delegate: Rc<dyn MessagePortDelegate>) {
+        match &mut *self.0.borrow_mut() {
+            MessagePortState::Ready { delegate, .. } => *delegate = Some(new_delegate),
+            MessagePortState::Detached => {},
+        }
+    }
+
+    pub fn delegate(&self) -> Option<Rc<dyn MessagePortDelegate>> {
+        match &*self.0.borrow() {
+            MessagePortState::Ready { delegate, .. } => delegate.clone(),
+            MessagePortState::Detached => None,
+        }
+    }
+
+    /// Prepare this MessagePort to be transferred as part of a PostMessage operation.
+    /// This object will no longer be usable afterwards.
+    pub fn into_jsvalue(&self) -> Result<JSValue, MessagePortDetached> {
+        let value = {
+            let state = self.0.borrow();
+            let &MessagePortState::Ready {
+                webview_id,
+                id,
+                entangled_id,
+                ..
+            } = &*state
+            else {
+                return Err(MessagePortDetached);
+            };
+            JSValue::MessagePort {
```

Comment:
The only place where we construct a `JSValue::MessagePort` is this method, but `into_jsvalue` is only called from unit tests. Am I missing something? Should `jsvalue_to_webdriver` also call this?